lib/trino-hive-formats/src/main/java/io/trino/hive/formats/ion/IonDecoderFactory.java

```diff
@@ -149,13 +150,52 @@ private static BlockDecoder decoderForType(Type type)
     private static BlockDecoder wrapDecoder(BlockDecoder decoder, Type trinoType, IonType... allowedTypes)
     {
         Set<IonType> allowedWithNull = new HashSet<>(Arrays.asList(allowedTypes));
+        return createConfigurableDecoder(decoder, trinoType, false, allowedTypes);
+    }
+
+    /**
+     * Wraps decoders for common handling logic.
+     * <p>
+     * Handles un-typed and correctly typed null values.
+     * Throws for mistyped values, whether null or not.
+     * Delegates to Decoder for correctly-typed, non-null values.
+     * Handles text coercion for Varchar and Char types.
+     * <p>
+     * This code treats all values as nullable.
+     */
+    private static BlockDecoder wrapDecoderWithTextCoercion(BlockDecoder decoder, Type trinoType, IonType... allowedTypes)
+    {
+        return createConfigurableDecoder(decoder, trinoType, true, allowedTypes);
+    }
+
+    /**
+     * Wraps decoders for common handling logic.
+     * <p>
+     * Handles un-typed and correctly typed null values.
+     * Throws for mistyped values, whether null or not.
+     * Delegates to Decoder for correctly-typed, non-null values.
+     * Handles text coercion for Varchar and Char types.
+     * <p>
+     * This code treats all values as nullable.
+     */
+    private static BlockDecoder createConfigurableDecoder(BlockDecoder decoder, Type trinoType, boolean textCoercion,
+                                                          IonType... allowedTypes)
+    {
+        final Set<IonType> allowedWithNull = new HashSet<>(Arrays.asList(allowedTypes));
         allowedWithNull.add(IonType.NULL);
 
         return (reader, builder) -> {
             final IonType ionType = reader.getType();
             if (!allowedWithNull.contains(ionType)) {
-                throw new TrinoException(StandardErrorCode.GENERIC_USER_ERROR,
-                        "Cannot coerce IonType %s to Trino type %s".formatted(ionType, trinoType));
+                if (textCoercion) {
```

Comment:
I'd prefer to avoid embedding a conditional that runs for each element into all of the non-text decoders.

Did you consider making it so that the Char and VarChar decoders simply accept any type and coerce there as needed? Is there a downside to that?

Reply:
Sure. I think I was trying to avoid specifying all `IonType`s in the allowed list but we have something called `IonType.values()` which gives all Ion type variants. Removed using the flag `textCoercion` and instead added all types as allowed types for `VarChar` and `Char` decoders.